govcd/catalog.go

```diff
@@ -55,3 +76,352 @@ func (c *Catalog) FindCatalogItem(catalogitem string) (CatalogItem, error) {
+}
+
+func uploadMultiPartFile(client *Client, filePaths []string, uploadHREF string, totalBytesToUpload int64) error {
+	log.Printf("[TRACE] Upload multi part file: %v\n, href: %s, size: %v", filePaths, uploadHREF, totalBytesToUpload)
+
+	var uploadedBytes int64
+
+	for i, filePath := range filePaths {
+		log.Printf("[TRACE] Uploading file: %v\n", i+1)
+		tempVar, err := uploadFile(client, uploadHREF, filePath, uploadedBytes, totalBytesToUpload)
+		if err != nil {
+			return err
+		}
+		uploadedBytes += tempVar
+	}
+	return nil
+}
+
+/** Function waits until vCD provides temporary file upload links. */
+func waitForTempUploadLinks(client *Client, vappTemplateUrl *url.URL) (*types.VAppTemplate, error) {
+	var vAppTemplate *types.VAppTemplate
+	var err error
+	for {
+		log.Printf("[TRACE] Sleep... for 5 seconds.\n")
+		time.Sleep(time.Second * 5)
+		vAppTemplate, err = queryVappTemplate(client, vappTemplateUrl)
+		if err != nil {
+			return nil, err
+		}
+		if len(vAppTemplate.Files.File) > 1 {
+			log.Printf("[TRACE] upload link prepared.\n")
+			break
+		}
+	}
+	return vAppTemplate, nil
+}
+
+func getOvfUploadLink(vappTemplate *types.VAppTemplate) (*url.URL, error) {
+	log.Printf("[TRACE] Parsing ofv upload link: %#v\n", vappTemplate)
+
+	ovfUploadHref, err := url.ParseRequestURI(vappTemplate.Files.File[0].Link[0].HREF)
+	if err != nil {
+		return nil, err
+	}
+
+	return ovfUploadHref, nil
+}
+
+func queryVappTemplate(client *Client, vappTemplateUrl *url.URL) (*types.VAppTemplate, error) {
+	log.Printf("[TRACE] Qeurying vapp template: %s\n", vappTemplateUrl)
+	request := client.NewRequest(map[string]string{}, "GET", *vappTemplateUrl, nil)
+	response, err := client.Http.Do(request)
+	if err != nil {
+		return nil, err
+	}
+	defer response.Body.Close()
+
+	body, err := ioutil.ReadAll(response.Body)
+	if err != nil {
+		return nil, err
+	}
+
+	vappTemplateParsed := &types.VAppTemplate{}
+
+	err = xml.Unmarshal(body, &vappTemplateParsed)
+	if err != nil {
+		return nil, err
+	}
+
+	log.Printf("[TRACE] Response: %v\n", response)
+	log.Printf("[TRACE] Response body: %s\n", string(body[:]))
+	log.Printf("[TRACE] Response body: %v\n", vappTemplateParsed)
+	return vappTemplateParsed, nil
+}
+
+/** Uploads ovf description file from unarchived provided ova file. As result vCD will generate temporary upload links which has to be queried later.
+Function will return parsed part for upload files from description xml.*/
+func uploadOvfDescription(client *Client, ovfFile string, ovfUploadUrl *url.URL) (Envelope, error) {
+	log.Printf("[TRACE] Uploding ovf description with file: %s and url: %s\n", ovfFile, ovfUploadUrl)
+	openedFile, err := os.Open(ovfFile)
+	if err != nil {
+		return Envelope{}, err
+	}
+
+	var buf bytes.Buffer
+	ovfReader := io.TeeReader(openedFile, &buf)
+
+	request := client.NewRequest(map[string]string{}, "PUT", *ovfUploadUrl, ovfReader)
+	request.Header.Add("Content-Type", "text/xml")
+
+	response, err := client.Http.Do(request)
```

Comment:
The rest of the library uses a function from api.go
```response, err := checkResp(client.Http.Do(request))```

And likewise, the response is unmarshaled using ```err = decodeBody(response, destinationObject)```
This feature should use the same approach, as it allows us to handle errors from a central place.

Reply:
Fixed